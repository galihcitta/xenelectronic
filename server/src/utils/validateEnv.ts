import { port, cleanEnv } from 'envalid';

export const validateEnv = () => {
  cleanEnv(process.env, {
    PORT: port()
  });
};

// https://www.npmjs.com/package/envalid
