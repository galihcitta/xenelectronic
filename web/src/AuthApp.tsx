import React from 'react';
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Redirect
} from 'react-router-dom';

import './index.css';

import { Layout } from './components/Layout/Layout.component';
import { Home } from './pages/Home/Home.component';
import { Shop } from './pages/Shop/Shop.component';
import { Cart } from './pages/Cart/Cart.component';
import { Details } from './pages/Details/Details.component';
import { NotFound } from './pages/NotFound/NotFound.component';

export const AuthApp: React.FC = () => (
  <Router>
    <Layout>
      <Switch>
        <Route exact path="/" component={Home} />
        <Route exact path="/shop" component={Shop} />
        <Route exact path="/cart" component={Cart} />
        <Route exact path="/product/:id" component={Details} />
        <Route path="/account" render={() => <Redirect to="/shop" />} />
        <Route component={NotFound} />
      </Switch>
    </Layout>
  </Router>
);

export default AuthApp;
