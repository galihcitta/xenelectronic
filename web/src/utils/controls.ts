import { Controls } from '../pages/Auth/types';

export const signInControls: Controls = {
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signin-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signin-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};

export const signUpControls: Controls = {
  username: {
    label: 'Enter username',
    placeholder: '',
    type: 'text',
    id: 'signup-username',
    name: 'username',
    value: '',
    validationRules: {
      required: true,
      minLength: 4,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signup-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signup-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true,
      isSecure: true,
      minLength: 8,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};
