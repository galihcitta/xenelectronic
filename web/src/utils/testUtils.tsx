import React from 'react';
import { BrowserRouter, Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../modules/store';
import { createMemoryHistory } from 'history';

const history = createMemoryHistory();

export const withRouter = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <BrowserRouter>
    <Component {...props} />
  </BrowserRouter>
);

export const withHistory = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Router history={history}>
    <Component {...props} />
  </Router>
);

export const withProvider = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Provider store={store}>
    <BrowserRouter>
      <Component {...props} />
    </BrowserRouter>
  </Provider>
);
