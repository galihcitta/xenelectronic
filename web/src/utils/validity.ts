import { ValidationRules } from '../pages/Auth/types';

export const validate = (
  value: string,
  controlName: string,
  rules: ValidationRules
) => {
  let isValid = true;
  let errorMsg = '';

  if (rules.required) {
    isValid = value.trim() !== '' && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} field is required`;
  }

  if (rules.minLength) {
    isValid = value.length >= rules.minLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be at least ${rules.minLength} characters`;
  }

  if (rules.maxLength) {
    isValid = value.length <= rules.maxLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be no longer than ${rules.maxLength} characters`;
  }

  if (rules.isEmail) {
    const pattern = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `email is invalid`;
  }

  if (rules.isSecure) {
    // Minimum eight characters, at least one uppercase letter, one lowercase letter, one number and one special character
    const pattern = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `password too weak`;
  }

  return {
    isValid,
    errorMsg: errorMsg.charAt(0).toUpperCase() + errorMsg.slice(1)
  };
};
