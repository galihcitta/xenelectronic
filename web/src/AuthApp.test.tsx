import { render } from '@testing-library/react';
import { AuthApp } from './AuthApp';
import { withProvider } from './utils/testUtils';
import { createMemoryHistory } from 'history';

const AuthAppWithProvider = withProvider(AuthApp);

// https://testing-library.com/docs/example-react-router
describe('<AuthApp />', () => {
  it('renders correctly', () => {
    const { asFragment } = render(AuthAppWithProvider({}));
    const history = createMemoryHistory();
    history.push('/account');

    expect(asFragment()).toMatchSnapshot();
  });
});
