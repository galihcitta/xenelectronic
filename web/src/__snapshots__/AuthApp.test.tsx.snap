// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthApp /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa EXKLM"
  >
    <div
      class="sc-chPdSV dmMliT"
      data-testid="Overlay"
    />
    <nav
      class="sc-jKJlTe iyDksZ"
      data-testid="Wrapper"
    >
      <header
        class="sc-eNQAEJ HwinD"
      >
        <p
          class="sc-hMqMXs lbCkvZ"
        >
          Interior & Exterior
        </p>
        <button
          aria-label="close"
          class="sc-kEYyzF kQBxvK"
          title="close menu"
        >
          <i
            class="far fa-times-circle"
          />
        </button>
      </header>
      <ul
        class="sc-kkGfuU kEahKy"
      >
        <li
          class="sc-kgoBCf eGQXyo"
        >
          <a
            class="sc-kpOJdX cDLtjk"
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="sc-kgoBCf eGQXyo"
        >
          <button
            class="sc-kGXeez gNIVWe"
          >
            Interior
            <i
              class="fas fa-angle-right"
            />
          </button>
        </li>
        <li
          class="sc-kgoBCf eGQXyo"
        >
          <button
            class="sc-kGXeez gNIVWe"
          >
            Exterior
            <i
              class="fas fa-angle-right"
            />
          </button>
        </li>
        <li
          class="sc-kgoBCf eGQXyo"
        >
          <a
            class="sc-kpOJdX cDLtjk"
            href="/account"
          >
            Account
          </a>
        </li>
        <li
          class="sc-kgoBCf eGQXyo"
        >
          <a
            class="sc-kpOJdX cDLtjk"
            href="/cart"
          >
            Cart (0)
          </a>
        </li>
      </ul>
    </nav>
    <nav
      class="sc-jKJlTe hcWqjD"
      data-testid="Wrapper"
    >
      <header
        class="sc-eNQAEJ HwinD"
      >
        <p
          class="sc-hMqMXs lbCkvZ"
        >
          Interior
        </p>
        <button
          aria-label="close"
          class="sc-kEYyzF kQBxvK"
          title="close menu"
        >
          <i
            class="far fa-times-circle"
          />
        </button>
      </header>
      <ul
        class="sc-kkGfuU kEahKy"
      >
        <li
          class="sc-kgoBCf eGQXyo"
        >
          <a
            class="sc-ckVGcZ fVtwdb"
            href="/shop"
          >
            Chairs
          </a>
        </li>
        <li
          class="sc-kgoBCf eGQXyo"
        >
          <a
            class="sc-ckVGcZ fVtwdb"
            href="/shop"
          >
            Lamps
          </a>
        </li>
        <li
          class="sc-kgoBCf eGQXyo"
        >
          <a
            class="sc-ckVGcZ fVtwdb"
            href="/shop"
          >
            Clocks
          </a>
        </li>
        <li
          class="sc-kgoBCf eGQXyo"
        >
          <a
            class="sc-ckVGcZ fVtwdb"
            href="/shop"
          >
            Vases
          </a>
        </li>
      </ul>
    </nav>
    <nav
      class="sc-jKJlTe hcWqjD"
      data-testid="Wrapper"
    >
      <header
        class="sc-eNQAEJ HwinD"
      >
        <p
          class="sc-hMqMXs lbCkvZ"
        >
          Exterior
        </p>
        <button
          aria-label="close"
          class="sc-kEYyzF kQBxvK"
          title="close menu"
        >
          <i
            class="far fa-times-circle"
          />
        </button>
      </header>
      <ul
        class="sc-kkGfuU kEahKy"
      >
        <li
          class="sc-kgoBCf eGQXyo"
        >
          <a
            class="sc-ckVGcZ fVtwdb"
            href="/shop"
          >
            Benches
          </a>
        </li>
        <li
          class="sc-kgoBCf eGQXyo"
        >
          <a
            class="sc-ckVGcZ fVtwdb"
            href="/shop"
          >
            Flowerpots
          </a>
        </li>
        <li
          class="sc-kgoBCf eGQXyo"
        >
          <a
            class="sc-ckVGcZ fVtwdb"
            href="/shop"
          >
            Plants
          </a>
        </li>
      </ul>
    </nav>
    <header
      class="sc-iwsKbI dYjCnF"
    >
      <div
        class="sc-gZMcBi fFtCwl"
      >
        <a
          class="sc-gqjmRU bmRGVH"
          href="/"
        >
          <h1
            class="sc-VigVT laIPOj"
          >
            Interior & Exterior
          </h1>
        </a>
        <nav
          class="sc-jTzLTM keBASo"
        >
          <button
            aria-label="open menu"
            class="sc-fjdhpX kBCHAW"
            title="Menu"
          >
            <i
              class="fas fa-bars"
            />
          </button>
          <ul
            class="sc-htpNat ezDfsx"
          >
            <li
              class="sc-EHOje gEpkr"
            >
              <a
                class="sc-ifAKCX ewhkVI"
                href="/shop"
              >
                Products
                <i
                  class="fas fa-caret-down"
                />
              </a>
              <div
                class="sc-bxivhb jmrBoK"
              >
                <ul
                  class="sc-bZQynM EvJhB"
                >
                  <li
                    class="sc-gzVnrw zSxKB"
                  >
                    <a
                      class="sc-htoDjs bOuEYB"
                      href="/shop"
                    >
                      Chairs
                    </a>
                  </li>
                  <li
                    class="sc-gzVnrw zSxKB"
                  >
                    <a
                      class="sc-htoDjs bOuEYB"
                      href="/shop"
                    >
                      Lamps
                    </a>
                  </li>
                  <li
                    class="sc-gzVnrw zSxKB"
                  >
                    <a
                      class="sc-htoDjs bOuEYB"
                      href="/shop"
                    >
                      Clocks
                    </a>
                  </li>
                  <li
                    class="sc-gzVnrw zSxKB"
                  >
                    <a
                      class="sc-htoDjs bOuEYB"
                      href="/shop"
                    >
                      Vases
                    </a>
                  </li>
                  <li
                    class="sc-gzVnrw zSxKB"
                  >
                    <a
                      class="sc-htoDjs bOuEYB"
                      href="/shop"
                    >
                      Benches
                    </a>
                  </li>
                  <li
                    class="sc-gzVnrw zSxKB"
                  >
                    <a
                      class="sc-htoDjs bOuEYB"
                      href="/shop"
                    >
                      Flowerpots
                    </a>
                  </li>
                  <li
                    class="sc-gzVnrw zSxKB"
                  >
                    <a
                      class="sc-htoDjs bOuEYB"
                      href="/shop"
                    >
                      Plants
                    </a>
                  </li>
                </ul>
              </div>
            </li>
            <li
              class="sc-EHOje gEpkr"
            >
              <a
                class="sc-ifAKCX ewhkVI"
                href="/account"
              >
                Account
              </a>
            </li>
            <li
              class="sc-EHOje gEpkr"
            >
              <a
                class="sc-ifAKCX ewhkVI"
                href="/cart"
              >
                Cart (0)
              </a>
            </li>
          </ul>
          <button
            aria-label="submit searching"
            class="sc-jzJRlG jLIVEF"
            title="Search"
          >
            <i
              class="fas fa-search"
            />
          </button>
        </nav>
      </div>
    </header>
    <form
      class="sc-iAyFgw iospd"
    >
      <input
        class="sc-hSdWYo TzbWC"
        placeholder="Search for products ..."
        type="test"
        value=""
      />
      <button
        class="sc-eHgmQL eEUXub"
        title="submit"
        type="submit"
      >
        <i
          class="far fa-arrow-alt-circle-right"
        />
      </button>
    </form>
    <main
      class="sc-bwzfXH fRIfPz"
    >
      <section
        class="sc-cvbbAY kcqui"
      >
        <a
          class="sc-cMljjf cLAgEn"
          href="/shop"
        >
          <h2
            class="sc-jWBwVP jfKoaf"
          >
            Lamps
          </h2>
          <div
            class="sc-brqgnP jaXLdd"
          >
            <img
              alt=""
              class="sc-jAaTju kLJKcd"
              src="lamp.png"
            />
          </div>
        </a>
        <a
          class="sc-cMljjf cLAgEn"
          href="/shop"
        >
          <h2
            class="sc-jWBwVP jfKoaf"
          >
            Vases
          </h2>
          <div
            class="sc-brqgnP jaXLdd"
          >
            <img
              alt=""
              class="sc-jAaTju kLJKcd"
              src="vase.png"
            />
          </div>
        </a>
        <a
          class="sc-cMljjf cLAgEn"
          href="/shop"
        >
          <h2
            class="sc-jWBwVP jfKoaf"
          >
            Chairs
          </h2>
          <div
            class="sc-brqgnP jaXLdd"
          >
            <img
              alt=""
              class="sc-jAaTju kLJKcd"
              src="chair.png"
            />
          </div>
        </a>
        <a
          class="sc-cMljjf cLAgEn"
          href="/shop"
        >
          <h2
            class="sc-jWBwVP jfKoaf"
          >
            Clocks
          </h2>
          <div
            class="sc-brqgnP jaXLdd"
          >
            <img
              alt=""
              class="sc-jAaTju kLJKcd"
              src="clock.png"
            />
          </div>
        </a>
        <a
          class="sc-cMljjf cLAgEn"
          href="/shop"
        >
          <h2
            class="sc-jWBwVP jfKoaf"
          >
            Benches
          </h2>
          <div
            class="sc-brqgnP jaXLdd"
          >
            <img
              alt=""
              class="sc-jAaTju kLJKcd"
              src="bench.png"
            />
          </div>
        </a>
        <a
          class="sc-cMljjf cLAgEn"
          href="/shop"
        >
          <h2
            class="sc-jWBwVP jfKoaf"
          >
            Flowerpots
          </h2>
          <div
            class="sc-brqgnP jaXLdd"
          >
            <img
              alt=""
              class="sc-jAaTju kLJKcd"
              src="flowerpot.png"
            />
          </div>
        </a>
        <a
          class="sc-cMljjf cLAgEn"
          href="/shop"
        >
          <h2
            class="sc-jWBwVP jfKoaf"
          >
            Plants
          </h2>
          <div
            class="sc-brqgnP jaXLdd"
          >
            <img
              alt=""
              class="sc-jAaTju kLJKcd"
              src="plant.png"
            />
          </div>
        </a>
      </section>
    </main>
    <footer
      class="sc-cSHVUG imUMEn"
    >
      <p
        class="sc-kAzzGY bLXsfs"
      >
        - Interior & Exterior -
      </p>
    </footer>
  </div>
</DocumentFragment>
`;
